pg_temp/pg_temp.py

```diff
@@ -103,22 +83,98 @@ def __init__(self,
 
         """
         self.verbosity = verbosity
-        with check_user() as (run_as, user_name):
-            options = dict() if not options else options
-            self._setup(databases, retry, tincr, initdb, postgres, sock_dir,
-                        psql, createuser, run_as, user_name, dirname, options)
+        self.docker_prefix = None
+        self.docker_container = None
+        self.docker_img = docker_img
+        # check for a postgres install, or fallback to docker
+        self._get_docker_fallback(postgres)
+        self.pg_process = None
+        # we won't expose this yet
+        self.run_as = self._get_run_as_account(None)
+        self.current_user = pwd.getpwuid(os.geteuid()).pw_name
+        options = dict() if not options else options
+        self._setup(databases, retry, tincr, initdb, postgres, sock_dir,
+                    psql, createuser, dirname, options)
+
+    def _get_docker_fallback(self, postgres_exe):
+        if self.docker_img:
+            # already using docker
+            return
+        if postgres_exe != DEFAULT_POSTGRES:
+            # exe was specified explicitly so don't use a fallback
+            return
+        if not shutil.which(DEFAULT_POSTGRES):
+            has_docker = shutil.which(DEFAULT_DOCKER_EXE)
```

Comment:
Oooh this looks really nice.  I didn't know about `shutil.which` in python3; now I know it's the modern replacement for `distutils.spawn.find_executable`.

Would it be okay to have `DEFAULT_DOCKER_EXE` as just `docker` so that the docker in `$PATH` wins?  This random macos laptop doesn't have `/usr/bin/docker` but using `shutils.which('docker')` does find it, so maybe we can have `DEFAULT_DOCKER_EXE = 'docker'` and stuff will be more generic?

Reply:
Ah great catch. I actually did have it as just plain 'docker' but I changed it to '/usr/bin/docker' since while I was testing the fall through mechanism I was unsetting my PATH so that the 'postgres' executable wouldn't be found, and I forgot to change it back. I'll fix this